pkg/clockcache/cache.go

```diff
@@ -0,0 +1,201 @@
+	}
+	values = values[:len(keys)]
+	self.lock.Lock()
+	defer self.lock.Unlock()
+
+	for idx := range keys {
+		keys[idx], values[idx], _ = self.insert(keys[idx], values[idx])
+	}
+	return
+}
+
+func (self *Cache) insert(key interface{}, value interface{}) (canonicalKey interface{}, canonicalValue interface{}, inserted bool) {
+	elem, present := self.elements[key]
+	if present {
+		return elem.key, elem.Value, false
+	}
+
+	var insertLocation *Element
+	if len(self.storage) >= cap(self.storage) {
+		insertLocation = self.evict()
+		*insertLocation = Element{key: key, Value: value}
+	} else {
+		self.storage = append(self.storage, Element{key: key, Value: value})
+		insertLocation = &self.storage[len(self.storage)-1]
+	}
+
+	self.elements[key] = insertLocation
+	return key, value, true
+}
+
+func (self *Cache) evict() (insertPtr *Element) {
+	for {
+		insertLocation, evicted := self.tryEvict()
+		if evicted {
+			return insertLocation
+		}
+	}
+}
+
+func (self *Cache) tryEvict() (insertPtr *Element, evicted bool) {
+	if self.next >= len(self.storage) {
+		self.next = 0
+	}
+
+	evicted = false
+	reachedEnd := false
+	for !evicted && !reachedEnd {
+		elem := &self.storage[self.next]
+		if elem.used != 0 {
+			elem.used = 0
+		} else {
+			insertPtr = elem
+			key := elem.key
+			delete(self.elements, key)
+			evicted = true
+		}
+		self.next += 1
+		reachedEnd = self.next >= len(self.storage)
+	}
+
+	return
+}
+
+// tries to get a batch of keys and store the corresponding values is valuesOut
+// returns the number of keys that were actually found.
+// NOTE: this function does _not_ preserve the order of keys; the first numFound
+//       keys will be the keys whose values are present, while the remainder
+//       will be the keys not present in the cache
+func (self *Cache) GetValues(keys []interface{}, valuesOut []interface{}) (numFound int) {
+	if len(keys) != len(valuesOut) {
+		panic(fmt.Sprintf("keys and values are not the same len. %d keys, %d values", len(keys), len(valuesOut)))
+	}
+	valuesOut = valuesOut[:len(keys)]
+	n := len(keys)
+	idx := 0
+
+	self.lock.RLock()
+	defer self.lock.RUnlock()
+
+	for idx < n {
+		value, found := self.get(keys[idx])
+		if !found {
+			if n == 0 {
+				return 0
+			}
+			// no value found for key, swap the key with the last element, and shrink n
+			n -= 1
+			keys[n], keys[idx] = keys[idx], keys[n]
+			continue
+		}
+		valuesOut[idx] = value
+		idx += 1
+	}
+	return n
+}
+
+func (self *Cache) Get(key interface{}) (interface{}, bool) {
+	self.lock.RLock()
+	defer self.lock.RUnlock()
+	return self.get(key)
+}
+
+func (self *Cache) get(key interface{}) (interface{}, bool) {
+
+	elem, present := self.elements[key]
+	if !present {
+		return 0, false
+	}
+
+	if atomic.LoadUint32(&elem.used) == 0 {
+		atomic.StoreUint32(&elem.used, 1)
+	}
+
+	return elem.Value, true
+}
+
+func (self *Cache) Unmark(key string) bool {
```

Comment:
Wouldn't a Remove be more appropriate? What's the use-case for just an Unmark where you wouldn't want a remove?

Reply:
1. Testing 😋 
2. incremental unmarking: one of the big downsides to this kind of cache is the ~linear time it takes to evict. One thing that may help with that is incrementally unmarking elements during normal operation so that when it comes time to evict there's less to do. `Remove` is much to aggressive an operation for that.